lib/Basics/DebugRaceController.cpp

```diff
@@ -24,47 +24,59 @@
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
 
 #include "DebugRaceController.h"
+
 #include "ApplicationFeatures/ApplicationServer.h"
+#include "Basics/ScopeGuard.h"
 
 using namespace arangodb;
 using namespace arangodb::basics;
 
 static DebugRaceController Instance{};
 
-DebugRaceController::DebugRaceController() {}
-
 DebugRaceController& DebugRaceController::sharedInstance() { return Instance; }
 
 void DebugRaceController::reset() {
-  std::unique_lock<std::mutex> guard(_mutex);
+  {
+    auto guard = std::scoped_lock<std::mutex>(_mutex);
+    _data.clear();
+  }
   _condVariable.notify_all();
-  _data.clear();
-  _didTrigger = false;
 }
 
-std::vector<std::any> DebugRaceController::data() const {
-  std::unique_lock<std::mutex> guard(_mutex);
-  return _data;
+bool DebugRaceController::didTrigger(
+    std::unique_lock<std::mutex> const& guard,
+    std::size_t numberOfThreadsToWaitFor) const {
+  TRI_ASSERT(guard.owns_lock());
+  TRI_ASSERT(_data.size() <= numberOfThreadsToWaitFor);
+  return numberOfThreadsToWaitFor == _data.size();
 }
 
 auto DebugRaceController::waitForOthers(
     size_t numberOfThreadsToWaitFor, std::any myData,
-    arangodb::application_features::ApplicationServer const& server) -> bool {
-  std::unique_lock<std::mutex> guard(_mutex);
-  if (!_didTrigger) {
+    application_features::ApplicationServer const& server)
+    -> std::optional<std::vector<std::any>> {
+  auto notifyGuard =
+      scopeGuard([this]() noexcept { _condVariable.notify_all(); });
+  {
```

Comment:
nit: extra block scope here looks unnecessary, but shouldn't do any harm.

Reply:
Yes, I added it just to emphasize that the mutex is unlocked before the scope guard is executed.